_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+  ASSERT(evt);
+  ...
+}
+```
+
+or check sizes of buffers
+
+```c
+void read_packet(void *buf, size_t buf_len) {
+  ASSERT(buf_len >= sizeof(Packet));
+  memcpy(...)
+}
+
+```
+
+and much more. The benefits of asserting on function arguments include:
+
+- developers connected to a debugger very quickly find programmatic mistakes
+- removes the need to check for argument validity and error handling code in the
+  upper layers
+- reduces chances of system faults by catching out-of-bounds accesses and
+  invalid pointer dereferences.
+
+It's not common to see asserts on arguments in vendor and library code, since
+they aren't in the business of crashing devices, but it is sometimes
+configurable on mature software. One example of configurable asserting in
+library code is the `xTaskCreateStatic` function in FreeRTOS[^4].
+
+In this function, `configASSERT` is used to check for non-null pointers, and if
+`configASSERT` is configured to a no-op, the code will still check those values
+later in the function and return an error if they are invalid.
+
+### When undefined behavior is immenent
+
+```c
+typedef enum {
+  EventType_A,
+  EventType_B,
+} EventType;
+
+void handle_event(Event *evt) {
+  ...
+  switch (evt->type) {
+    case EventType_A:
+      // something
+    case EventType_B:
+      // something
+    default:
+      // shouldn't ever get here.
+      // something is corrupted
+      ASSERT(0);
+  }
+}
+```
+
+If we want to make this even more understandable for future developers who come
+across this, something like the following can also be done.
+
+```c
+#define WTF ASSERT(0)
+
+
+void handle_event(Event *evt) {
+  ...
+  switch (evt->type) {
+    ...
+    default:
+      WTF;
+  }
+}
+```
+
+This makes it pretty obvious that anything that falls through to the `default:`
+case is undefined behavior.
+
+### RTOS API functions fail
+
+Another reasonable place to assert is when you expect certain system function to
+succeed, and if they don't, the system is in a bad state. Below is a snippet
+where we assert if a mutex wasn't locked with a few seconds, which likely means
+there is a deadlock somewhere else.
+
+```c
+int rv = xSemaphoreTake(s_mutex, sec_to_ticks(5));
+ASSERT(rv == pdTRUE);
+```
+
+## When not to assert
+
+There are places where it is advised to assert. On the contrary, there are
+places where it's critical to **not** assert.
+
+- Avoid asserting on boot up sequences. This is the most common way to introduce
+  reboot loops.
+- Don't assert on operations that depend on the hardware behaving appropriately.
+  If a sensor says it will return a value between 0-100, it's probably best not
+  assert when it's above 100, because you can never trust today's cheap
+  hardware.
+- Don't assert on the contents of data read from persistent storage, unless it's
+  guaranteed to be valid. The data read from flash or a filesystem could be
+  corrupted.
+- If it is very likely the system will recover in a few moments, it might be
```

Comment:
Maybe an example here? What would the system be recovering from?

Reply:
I added your network back pressure. Was a good example.